addons/gamification/models/res_users.py

```diff
@@ -42,35 +64,38 @@ def _get_user_badge_level(self):
     @api.model_create_multi
     def create(self, values_list):
         res = super(Users, self).create(values_list)
-
-        karma_trackings = []
-        for user in res:
-            if user.karma:
-                karma_trackings.append({'user_id': user.id, 'old_value': 0, 'new_value': user.karma})
+        karma_trackings = [
+            {
+                'new_value': vals.get('karma'),
+                'old_value': 0,
+                'origin_ref': f'res.users,{self.env.uid}',
+                'reason': _('User Creation'),
+                'user_id': user.id,
+            }
+            for user, vals in zip(res, values_list)
+            if vals.get('karma')
+        ]
         if karma_trackings:
             self.env['gamification.karma.tracking'].sudo().create(karma_trackings)
 
-        res._recompute_rank()
         return res
 
-    def write(self, vals):
-        karma_trackings = []
-        if 'karma' in vals:
-            for user in self:
-                if user.karma != vals['karma']:
-                    karma_trackings.append({'user_id': user.id, 'old_value': user.karma, 'new_value': vals['karma']})
-
-        result = super(Users, self).write(vals)
+    def _add_karma(self, karma, source=None, reason=None):
```

Comment:
Quick proposal as we were discussing this with Thibault this morning.

In e-learning we might have a use case where we unlink 200 memberships at once to courses.
Each user will have earned a different amount of karma (so "_add_karma" will not work for this).

It would be ideal to have a method "_add_karma_batch" that lets you pass some kind of dict {user: {karma, source, reason}} that creates the tracking records in batch.
Don't hesitate to ask if that's unclear.

Cheers!

cc @tde-banana-odoo